File: Rimon/LeetCode/1382_Balance a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        
        st = []
        
        def inOrder(root,st):
            
            if root == None:
                return st 
            if root.left:
                st =  inOrder(root.left,st)
            st.append(root)
            if root.right:
                st = inOrder(root.right,st)
            return st
        def balanceTree(st):
            # print(st)
            if len(st)<=0:
                return None
            rt = st[len(st)//2]
            rt.left = balanceTree(st[:len(st)//2])
            rt.right = balanceTree(st[len(st)//2 +1 : ])
            return rt
        
        st = inOrder(root,st)
        # print(st)
        return balanceTree(st)
        

File: Rimon/LeetCode/1382_Balance a Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> st;
    void inOrder(TreeNode* root){
        if (root!=NULL){
        
            
            inOrder(root->left);
            st.push_back(root);
            inOrder(root->right); 
        }
        
    }
    TreeNode* createBST(int s,int e){
        if(s>e)
            return NULL;
        int mid = (s + e)/2;
        TreeNode *temp = st.at(mid);
        temp->left = createBST(s,mid-1);
        temp->right = createBST(mid+1,e);
        return temp;
    }
    TreeNode* balanceBST(TreeNode* root) {
        inOrder(root);
        int l = st.size();
        return createBST(0,l-1);
    }
};